File: modules/landing_zone_reader_config/scripts/apply.py
import os
import json
from libs import cli, execWithErrors, execWithoutErrors
from six import string_types

def main():
    components = eval(os.environ['COMPONENTS'])
    
    include = []
    for k in components.keys():
        include.append(k)
    return terrahubOutput(include)

def terrahubOutput(include):
    response = {}
    output_map = []

    execWithoutErrors(['terrahub', 'configure', '-c', 'component.template.data', '-D', '-y', '-i', 'terrahub_load_outputs'], os.environ['ROOT_PATH'])
    for include_item in include:
        result = ''
        (error, result) = cli(['terrahub', 'output', '-o', 'json', '-i', include_item, '-y'], os.environ['ROOT_PATH'])
        if error == 0:
            prepare_output = 'map(' + ','.join(extractOutputValues(result)) + ')'
            output_map = output_map + [prepare_output]

    (error, result) = cli(['terrahub', 'configure', '-i', 'terrahub_load_outputs', '-c', 'component.template.output.terrahub_reader.value=${merge(' + ','.join(output_map) + ')}'], os.environ['ROOT_PATH'])
    if error != 0:
        print('Error')
    (error, result) = cli(['terrahub', 'run', '-i', 'terrahub_load_outputs', '-a', '-y'], os.environ['ROOT_PATH'])
    if error != 0:
        print('Error')
    (error, result) = cli(['terrahub', 'refresh', '-i', 'terrahub_load_outputs'], os.environ['ROOT_PATH'])
    if error != 0:
        print('Error')

    return 'Success'

def extractOutputValues(result):
    response = {}
    processes = []
    thub_cfg = ['terrahub', 'configure', '-i', 'terrahub_load_outputs', '-c']

    output_map = []
    for (key, val) in json.loads(result).items():
        processes.append(thub_cfg + ['component.template.data.terraform_remote_state.'+ key + '.backend=local'])
        processes.append(thub_cfg + ['component.template.data.terraform_remote_state.'+ key + '.config.path=${tfvar.terrahub["tfstate_path"]}/' + key + '/terraform.tfstate'])
        for key_sub in val.keys():
            output_map = output_map + ['"' + key_sub + '"', 'data.terraform_remote_state.' + key + '.' + key_sub]
    
    execWithErrors(processes, os.environ['ROOT_PATH'])

    return output_map

if __name__ == '__main__':
    RESP = main()
    print(RESP)

File: modules/landing_zone_reader_config/scripts/apply.js
'use strict';

const Helper = require('./helper');

const { ROOT_PATH: rootPath, BACKEND: backend, COMPONENTS: components } = process.env;

/**
 * Check if required env variables are defined
 */
function checkEnvironmentVars() {
  if (!rootPath) {
    throw Error('ERROR: ROOT_PATH variable is empty. Aborting...');
  }

  if (!backend) {
    throw Error('ERROR: BACKEND variable is empty. Aborting...');
  }

  if (!components) {
    throw Error('ERROR: COMPONENTS variable is empty. Aborting...');
  }
}

/**
 * Run terrahub output
 * @param {Array} include
 * @param {JSON} jsonBackend
 */
async function terrahubOutput(include, jsonBackend) {
  let outputMap = [];

  Helper.executeWithoutErrors(
    rootPath, 'terrahub',
    [
      'configure', '--config', 'component.template.data', '--delete',
      '--auto-approve', '--include', 'terrahub_load_outputs'
    ]
  );

  await Promise.all(
    include.map(async item => {
      Helper.cli(
        rootPath, 'terrahub',
        ['init', '--include', item]
      );

      const result = Helper.cli(
        rootPath, 'terrahub',
        ['output', '--format', 'json', '--include', item, '--auto-approve']
      );

      if (!result.length) {
        throw new Error('No terraform outputs found. Before using `landing_zone_reader` module, ' +
          'make sure that `landing_zone` module generates output. Learn more: https://github.com/MitocGroup/terraform-aws-landing-zone/');
      }

      const outputValues = await extractOutputValues(result, jsonBackend);
      const prepareOutput = `map(${outputValues.join(',')})`;

      outputMap = [...outputMap, ...[prepareOutput]];
    })
  );


  Helper.cli(
    rootPath, 'terrahub',
    [
      'configure', '--include', 'terrahub_load_outputs', '--config',
      `component.template.output.terrahub_reader.value=merge(${outputMap.join(',')})`
    ]);


  Helper.cli(
    rootPath, 'terrahub',
    ['run', '--include', 'terrahub_load_outputs', '--apply', '--auto-approve']
  );

  return 'Success';
}

/**
 * Extract output values
 * @param {String} result
 * @param {JSON} jsonBackend
 * @return {Promise<Array>}
 */
async function extractOutputValues(result, jsonBackend) {
  const processes = [];
  let outputMap = [];
  const terrahubConfig = ['configure', '--include', 'terrahub_load_outputs', '--config'];
  const jsonResult = JSON.parse(result);

  Object.keys(jsonResult).forEach(key => {

    const jsonBackendKeysArray = Object.keys(jsonBackend);
    const { backend } = jsonBackend;
    processes.push([
      ...terrahubConfig,
      ...[
        `component.template.data.terraform_remote_state.${key}` +
        `.backend=${backend}`
      ]
    ]);
    processes.push([
      ...terrahubConfig,
      ...[
        `component.template.data.terraform_remote_state.${key}` +
        `.config={}`
      ]
    ]);
    jsonBackendKeysArray.filter(elem => elem !== 'backend').forEach(backendKey => {
      let backendValue = jsonBackend[backendKey];
      if (['key', 'path', 'prefix'].indexOf(backendKey) > -1) {
        backendValue += `/${key}` +
          (backend === 'prefix' ? '' : '/terraform.tfstate');
      }
      processes.push([
        ...terrahubConfig,
        ...[
          `component.template.data.terraform_remote_state.${key}` +
          `.config.${backendKey}=${backendValue}`
        ]
      ]);
    });

    Object.keys(jsonResult[key]).filter(elem => elem.includes('landing_zone') ).forEach(subKey => {
      outputMap = [...outputMap, ...[`"${subKey}"`, `data.terraform_remote_state.${key}.outputs.${subKey}`]];
    });
  });

  try {
    Helper.executeWithErrors(rootPath, 'terrahub', processes);
  } catch (error) {
    throw error;
  }

  return outputMap;
}

/**
 * Execute
 * @return {Promise}
 */
async function main() {
  const jsonBackend = JSON.parse(backend);
  const jsonComponents = JSON.parse(components);
  const include = [];

  Object.keys(jsonComponents).forEach(key => include.push(key));

  return terrahubOutput(include, jsonBackend);
}

(async () => {
  try {
    checkEnvironmentVars();
    Helper.checkIfTerrahubIsInstalled();
    const resp = await main();

    console.log(resp);
  } catch (error) {
    console.log(error);
  }
})();
